Implementation for r_set_equal()

matthiasbeyer opened the pull request:
This PR is a WIP for adding the implementation of `r_set_equal()` along with
the internal helpers and therefor targets #174 .

@neithernut I would like to have some pointers on how to approach this.
I don't think I have to do a complex operation where I loop in O(n^2)
trough all ht buckets and inside that in O(n^2) through all avl nodes,
and again, inside that one in O(n^2) through all linked lists? Seems
really much to complex for me to be the way how this has to be solved...

matthiasbeyer commented:
Should be rebased after #205 was merged.

TheNeikos commented on src/libreset/avl/avl_equal.c:
```diff
@@ -0,0 +1,26 @@
+    struct avl const* avlb,
+) {
+    if (avla == avlb ||
+        (avla == NULL && avlb != NULL) || (avla != NULL && avlb == NULL)) {
```
In the header of avl_equal you have this attributed as "nonnull". This condition doesn't make sense to me in this case.

matthiasbeyer replied:
I don't know if the compiler inserts checks whether the passed arguments are `NULL` when specifying this macro. @neithernut do you?

TheNeikos commented on src/libreset/avl/avl_equal.c:
```diff
@@ -0,0 +1,26 @@
+) {
+    if (avla == avlb ||
+        (avla == NULL && avlb != NULL) || (avla != NULL && avlb == NULL)) {
+        return 1;
```
How is an AVL that is non existent equal to one that does exist?

matthiasbeyer commented:
@TheNeikos removed your comment. Was not in the diff and anyway fixed by a fixup commit.

matthiasbeyer commented:
Apparently, I have no clue where to go from here. @neithernut ?

neithernut commented on src/libreset/avl/avl.h:
```diff
@@ -131,6 +131,21 @@ avl_find(
+int
+avl_equal(
+    struct avl const* avla, //!< The first avl object for the comparison
+    struct avl const* avlb, //!< The second avl object for the comparison
```
remove the `,`

neithernut commented on src/libreset/ll/ll_equal.c:
```diff
@@ -0,0 +1,28 @@
+    int res = 1;
+
+    ll_foreach(it_a, lla) {
+        ll_foreach(it_b, llb) {
```
This loop will not work as intended. Use `ll_find` instead.

neithernut replied:
Also: the intention of this double loop-construct is to check whether all elements of one set are in the other, too. This, however, doesn't neccessarily mean equality. It means nothing but A being the subset of B.
You either have to do it the other way around too (ugh), or verify that the number of elements are equal, too.

neithernut commented:
First thing you will need is some way to map buckets from one hashtable to buckets for another ht.
This is functionality needed for a few algorithms, so put it in `base.h`/`base.c`.
Next thing yo will need is some way to check the equality of one AVL with a few AVLs.
The latter one is not as trivial as it may seem, though it may be simplified. I'll spend a few hours of sleep on this issue.

neithernut commented on src/libreset/ll/ll_equal.c:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    if ((lla == NULL) ^ (llb == NULL)) {
+        return 1;
```
This should probably be `return 0;` instead.

matthiasbeyer commented on src/libreset/avl/avl_equal.c:
```diff
@@ -0,0 +1,28 @@
+    struct avl_el* roota,
+    struct avl_el* rootb
+) {
+
```
@neithernut Do you know some hacks how to implement this?

neithernut replied:
Well, nothing _nice_ yet.

TheNeikos replied:
What about `roota === rootb` it does type checking as well!

neithernut replied:
This does not check the contents.
Anyways, this will most certainly be implemented after the switch to done.io (PR #222), anyways.

matthiasbeyer replied:
Good joke!

matthiasbeyer commented:
@neithernut Please review.

neithernut commented:
ACK.

matthiasbeyer commented on src/libreset/avl/avl_is_subset.c:
```diff
@@ -0,0 +1,45 @@
+#include "avl.h"
+
+int
```
`static` and documentation missing for this function.

matthiasbeyer replied:
ASCII-Art welcome here!

matthiasbeyer commented on src/libreset/avl/avl_is_subset.c:
```diff
@@ -0,0 +1,45 @@
+    struct r_set_cfg const* cfg
+) {
+    if (!node_a) {
+        return 1;
```
Please explain why you return `1` here.

matthiasbeyer commented on src/libreset/avl/avl_is_subset.c:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    if (!node_b) {
+        return 0;
```
... and `0` here.

matthiasbeyer commented on src/libreset/avl/avl_is_subset.c:
```diff
@@ -0,0 +1,45 @@
+    struct r_set_cfg const* cfg
+) {
+    if (!avl_a) {
+        return 1;
```
Why do you return `1` here? (Or is this notes in the interface definition? Dunno.)

matthiasbeyer commented on src/libreset/ht/ht_equal.c:
```diff
@@ -0,0 +1,56 @@
+        for (int i = ht_b->sizeexp; i; --i) {
+            for (int j = denom_buckets; j; --j) {
+                if (!avl_is_subset(&ht_a->buckets[i+j].avl,
+                            &ht_b->buckets[i].avl, cfg)) {
```
I guess this line is one indentation level too much on the right.

matthiasbeyer commented on src/libreset/ht/ht_equal.c:
```diff
@@ -0,0 +1,56 @@
+    } else {
+        for (int i = ht_b->sizeexp; i; i--) {
+            if (!avl_is_subset(&ht_a->buckets[i].avl,
+                        &ht_b->buckets[i].avl, cfg)) {
```
Also here: indentation level.

matthiasbeyer commented on src/libreset/ll/ll.h:
```diff
@@ -114,6 +114,22 @@ __r_warn_unused_result__
+ *
+ * @memberof ll
+ *
+ * @return 0 if ll_a is a subset of ll_b else a 1
```
Are the return values switched here? `is_subset` is a boolean operation and should therefor return `1` on success and `0` on failure.

matthiasbeyer commented on src/libreset/set.c:
```diff
@@ -114,3 +114,19 @@ r_set_cardinality(
+    struct r_set const* set_a,
+    struct r_set const* set_b
+) {
+    if (set_a == set_b || !set_a || !set_b) {
```
If `set_a` or `set_b` are `NULL`, you return `1`, which indicates that they are equal?

matthiasbeyer commented on tests/avl/avl_tests.c:
```diff
@@ -254,6 +293,7 @@ suite_avl_create(void) {
     case_adding     = tcase_create("Adding");
     case_deleting   = tcase_create("Deleting");
     case_finding    = tcase_create("Finding");
+    case_subset    = tcase_create("Finding");
```
Indentation and name wrong here.

matthiasbeyer commented on tests/ht/ht_tests.c:
```diff
@@ -173,12 +251,18 @@ suite_ht_create(void) {
+    tcase_add_test(case_equality, test_ht_equal_wrong);
+    tcase_add_test(case_equality, test_ht_equal_almost);
+    tcase_add_test(case_equality, test_ht_equal_different_buckets);
+
```
Test missing for different bucket sizes and non-equal hashtables.

matthiasbeyer commented on tests/ht/ht_tests.c:
```diff
@@ -141,6 +141,82 @@ START_TEST (test_ht_cardinality) {
+
+    ck_assert(ht_equal(&ht, &ht2, &cfg_int) == 1);
+
+    ck_assert(0 == ht_destroy(&ht, &cfg_int));
```
You can outsource most code of the above 3 tests into a (static) helper function and just pass the data and the expected return value.

TheNeikos replied:
Is it necessary?

matthiasbeyer replied:
It would safe you _a lot_ of code and therefor a lot of maintaining work for you and others would be saved...

matthiasbeyer commented on tests/ll/ll_test.c:
```diff
@@ -147,12 +147,68 @@ START_TEST (test_ll_count) {
+    struct ll* ll2 = calloc(1, sizeof(*ll2));
+    int i;
+    int data[] = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 };
+    int data2[] = { 0, 1, 2, 3, 4, 5, 6, 8, 9, 10 };
```
Can you please change this line to `... 5, 6,     8, 9, 10`, so you can simply visually notice that the data is different here.

TheNeikos replied:
But that is what is written there, isn't it?

matthiasbeyer replied:
Nah, github erased the whitespace.

I meant that you should increase the whitespace between the `6` and the `8`, so one can easily recognize the difference of the data arrays.

TheNeikos replied:
Alright

matthiasbeyer commented on tests/set/set_tests.c:
```diff
@@ -13,21 +13,42 @@ START_TEST (test_r_set_cardinality) {
 
     /* Test case creation */
     case_cardinality  = tcase_create("Cardinality");
+    case_equality  = tcase_create("Equality");
```
Indentation wrong here.

matthiasbeyer commented:
@neithernut Are you fucking serious?

neithernut commented:
The core algorithms were sound and I did not consider the documentation too sparse.
But then again, maybe the fact that I was sitting right next to him and discussing the algorithms with him made my view a little superficial regarding the documentation.

matthiasbeyer commented:
@TheNeikos I don't know how much is left to do here (is there anything left at all?). But I would highly appreciate a shorter history for this PR. So if you can squash some commits when you're done with the code, that would be really fine!

And, of course, the remaining comments in the diff. I removed everything else which was discussed and rejected by you, so we keep the PR clear.

TheNeikos committed:
Implemented ll_is_subset

TheNeikos committed:
Implemented ll_equal

TheNeikos committed:
Implemented avl_is_subset

TheNeikos committed:
Implemented ht_equal

TheNeikos committed:
Implemented r_set_equal

TheNeikos committed:
Added files to CMakeLists

TheNeikos committed:
Added avl_is_subset tests

TheNeikos committed:
Added ht_equal tests

TheNeikos committed:
Added ll_equal and ll_is_subset tests

TheNeikos committed:
Added r_set_equal tests

TheNeikos committed:
Changed the test config

This commit changes the hashf so that the hash as quite larger hashes.
This way the hashes are more randomly distributed which is great for testing.

TheNeikos committed:
Fixed missing const modifier

matthiasbeyer commented:
@neithernut Please review.

neithernut commented:
ACK.
